File: ex414_third_max_num/solution.cpp
class Solution {
public:
    // Walk the list and basically attempt perform comparisons for each
    int thirdMax(vector<int>& nums) {
        int64_t largest, middle, third, insertCount(0);
        largest = middle = third = std::numeric_limits<int64_t>::min();
                    
        for (auto num : nums) {
            if (num == largest || num == middle || num == third)  
                continue;
            
            if (num > largest) {
                third = middle;
                middle = largest;
                largest = num;
                insertCount++;
            } else if (num > middle) {
                third = middle;
                middle = num;
                insertCount++;
            } else if (num > third) {
                third = num;
                insertCount++;
            }
        }
        return (insertCount < 3) ? largest : third;
    }
};

File: ex414_third_max_num/solution.py
class Solution(object):
    def thirdMax(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        deduped_list = list(sorted(set(nums)))
        if len(deduped_list) < 3:
            return deduped_list[-1];
        
        return deduped_list[-3];
        
